pkg/providers/cloudstack/cloudstack.go

```diff
@@ -641,6 +641,9 @@ func (cs *CloudStackTemplateBuilder) GenerateCAPISpecControlPlane(clusterSpec *c
 	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
 		etcdMachineSpec = *cs.etcdMachineSpec
 	}
+	if clusterSpec.CloudStackDatacenter == nil {
+		return nil, fmt.Errorf("provided clusterSpec CloudStackDatacenter is nil. Unable to build template map CP")
```

Comment:
If this is a concerning scenario, we should probably have this as part of the validation layer, not here.